File: Heaps/MedianFinder.java
package Heaps;

import java.util.Comparator;
import java.util.PriorityQueue;

public class MedianFinder {
    private PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Comparator.reverseOrder()); // stores smaller half
    private PriorityQueue<Integer> minHeap = new PriorityQueue<>(); // stores larger half

    public void addNum(int num) {
        maxHeap.add(num);
        minHeap.add(maxHeap.remove());
        if (minHeap.size() > maxHeap.size()) {
            maxHeap.add(minHeap.remove());
        }

    }

    public double findMedian() {
        if (maxHeap.size() > minHeap.size()) {
            return maxHeap.peek();
        }

        return (maxHeap.peek() + minHeap.peek()) / 2.0;
    }
}


// This algorithm allows us to have an O(1) time complexity for findMedian 
// and an  O(logn) time complexity for addNum, which makes it an incredibly fast algorithm,
// where n is the number of times addNum has been called so far. The space complexity is 
// O(n) to store the heaps.

File: Heaps/median_finder.py
from heapq import heappop, heappop_max, heappush, heappush_max


class MedianFinder:
    def __init__(self):
        self.min_heap = []
        # Max heap will store median when we have odd count of ints
        self.max_heap = []

    def addNum(self, num: int) -> None:
        # Max heap will store the smaller elements
        # Min heap will store the larger elements
        heappush_max(self.max_heap, num)
        heappush(self.min_heap, heappop_max(self.max_heap))
        if len(self.min_heap) > len(self.max_heap):
            heappush_max(self.max_heap, heappop(self.min_heap))

    # Time complexity: O(log n)

    def findMedian(self) -> float:
        if len(self.max_heap) > len(self.min_heap):
            return self.max_heap[0]
        return (self.max_heap[0] + self.min_heap[0]) / 2

    # Time complexity: O(1)
